Add generator docs (#207)
kasiopea-org/pisek: docs/generator.md

# Generator
The generator is used for generating inputs. Solutions are then run and judged on those.

## Generator format
There are currently 3 generator formats available:
- pisek-gen-v1
- cms-old
- opendata-v1

However we strongly recommend using the pisek-gen format
for better debugging and easy conversion between open data and closed data tasks.

## Terminology
There are two requirements for generators:
- Generators **must** be deterministic. (For the same arguments, it should generate the same input(s).) 
- If a generator takes a seed as an argument, the generator **should** respect the seed. (For different seeds
  it should generate different inputs.) This can be disabled in the `[checks]` section, but be careful.

## Pisek-gen-v1
### Listing inputs
When run without arguments it should list all inputs it can generate in the following format: 
```
input_name key1=value1 key2=value2
```
Where `input_name` is the name of the given input. The input will be generated into the file
`[input_name]_[seed].in` or `[input_name]` (depending whether the input is seeded).
This is followed by any number of key=value pairs separated by spaces.
The following keys are supported:

> | Key    | Meaning                                       | Value type | Default value |
> | ------ | --------------------------------------------- | ---------- | ------------- |
> | repeat | How many times should this input be generated? | int        | 1             |
> | seeded | Is this input generated depending on a seed?   | bool       | true          | 

Reviewer: ```suggestion
| Key    | Meaning                                        | Value type | Default value |
| ------ | ---------------------------------------------- | ---------- | ------------- |
| repeat | How many times should this input be generated? | int        | 1             |
| seeded | Is this input generated with a random seed?    | bool       | true          | 
```

If input is seeded, repeat must be 1.

For example:
```
01_tree
02_random_graph repeat=10
02_complete_graph seeded=false
```

### Generating inputs

> Generated is then repeatedly asked to generated input `input_name` from
> inputs list.

Reviewer: ```suggestion
The generator is then repeatedly asked to generated the input `input_name` from
the inputs list.
```

> If `input_name` is seeded generator is run:

Reviewer: ```suggestion
If `input_name` is seeded the generator is run with:
```

```
./gen [input_name] [seed]
```

> Where `seed` is a hexadecimal number. Generator must be deterministic and
> respect seed.

Reviewer: ```suggestion
Where `seed` is a hexadecimal number. The generator must be deterministic and
respect the seed.
```

> If `input_name` is unseeded generator is  

Reviewer: ```suggestion
If `input_name` is unseeded the generator is called with  
```

```
./gen [input_name]
```

> Generator must be deterministic.

Reviewer: ```suggestion
The generator must be deterministic.
```

> In either case, generator should print the input to it's stdout. 

Reviewer: ```suggestion
In either case, the generator should print the input to its stdout. 
```

## Cms-old

> Generator is run:

Reviewer: ```suggestion
The generator is run with:
```

```
./gen [directory]
```

> Generator should generate all input files to this directory. Generator must be deterministic.

Reviewer: ```suggestion
The generator should generate all input files to this directory. The generator must be deterministic.
```

## Opendata-v1

> Generator is run:

Reviewer: ```suggestion
The generator is run:
```

```
./gen [subtask] [seed]
```

> Generator should generate input for this subtask to it's stdout. Generator must be deterministic
> and respect given seed.

Reviewer: ```suggestion
The fenerator should generate the input for this subtask to its stdout. The generator must be deterministic
and respect the given seed.
```

> (Please note that generator can generate only one input for each subtask.)

Reviewer: ```suggestion
(Please note that the generator can generate only one input for each subtask.)
```
